Add first implementation of htcondor collector.

rfvc opened the pull request:
Hey all,

this implements a collector for collecting data from an HTCondor batch system. It uses the `pyauditor`-client for connecting to an AUDITOR-instance. The collector is currently collecting data at the German tier 1 WLCG site.

Cheers, 
  Florian

stefan-k reviewed (changes requested):
Thanks for this contribution! :)

I'm on vacation, therefore I'm only able to give a _ReviewLight_ ™. 
I've never seen such a red CI, but that's not your fault. The Rust part is organized in so-called _workspaces_, and it is configured to treat everything in `/collectors/` as a Rust crate. For this to work, you will have to change the `Cargo.toml` in the project's root to the following:

```toml
[workspace]
members = [
    "auditor",
    "pyauditor",
    "collectors/slurm",
    "collectors/slurm-epilog",
    "plugins/priority",
]

[profile.release]
strip = true
```

There is another point which will probably require more discussion: It would be great to have integration tests which test this collector against the current version of Auditor and pyauditor, similarly as it is done [here](https://github.com/ALU-Schumacher/AUDITOR/blob/main/scripts/test_slurm_collector.sh) and [here](https://github.com/ALU-Schumacher/AUDITOR/blob/main/scripts/test_slurm_epilog_collector.sh). This will require building the pyauditor Python package and spinning up an HTCondor cluster in a container. This should be run in the CI. I think we need to discuss the details and whether this should be part of this PR or a dedicated PR. 
 

I will give a more in-depth review once I'm back in the office. For the time being I'll assign our Python experts @dirksammel  and @QuantumDancer to review the Python code.

dirksammel commented on collectors/htcondor/collector.py:
```diff
@@ -0,0 +1,242 @@
+    def _generate_components(self, job: dict) -> List[Component]:
+        components = []
+        for component in self.config.components:
+            if amount := get_value(component, job):
```
See above.

dirksammel commented on collectors/htcondor/collector.py:
```diff
@@ -0,0 +1,242 @@
+        for key, entry in self.config.meta.items():
+            values = []
+            for item in entry if isinstance(entry, list) else [entry]:
+                if (value := get_value(item, job)) is not None:
```
See above.

dirksammel commented on collectors/htcondor/collector.py:
```diff
@@ -0,0 +1,242 @@
+                        break
+            if not values:
+                self.logger.warning(
+                    f"Could not find meta value for \"{key}\" for job \"{job['GlobalJobId']}\"."
```
Here you escape `"`, in other log messages you start the f-string with `'`, maybe try do this in a consistent style.

maxfischer2781 replied:
Consider to use raw formatting as `... for {key!r} for ...` (no quotes, but `!r` instead). This will automatically add quotes but also escape special characters that would otherwise be invisible.

rfvc replied:
You are right, I've changed that to use raw strings for logging.

rfvc commented on collectors/htcondor/README.md:
Done.

rfvc commented:
@stefan-k, thanks for the ReviewLight™ 😅 I've moved the docu to the website and changed the `Cargo.toml` to explicitly state the slurm-collectors so the CI should hopefully become greener again.

I'll have a look at how to add tests for the collector in the CI.

Enjoy your vacation!

QuantumDancer reviewed (commented):
Apart from a few minor comments, I have a general one: in `collector.py` there are type annotations, which I find pretty nice. However, in other files, there are either no type hints (e.g. `state_db.py`) or the type information is only given in doc strings (e.g. `utils.py`). Would it make sense to also add type annotations in these files?

maxfischer2781 commented on collectors/htcondor/collector.py:
```diff
@@ -0,0 +1,242 @@
+
+        added, failed = 0, 0
+        for job in jobs[::-1]:
+            if record := self._generate_record(job):
```
Support for Python 3.6 (!) would be great, since that is still the default on RHEL7 run by many grid sites.

This would be easier if `_generate_record` would raise an exception if no record can be generated.

maxfischer2781 replied:
Scratch 3.6. I only checked the `pyproject.toml` but the Maturin build backend already dropped support for 3.6. Targeting 3.8 should be fine, then. The distributions I'm aware of directly went from 3.6 to 3.8, plus 3.7 will be EOL soon as pointed out already.

I'd still recommend handling this via exceptions instead of `None` (at the very least, check for `is None`) but in principle both is fine.

rfvc replied:
Thanks @dirksammel and @maxfischer2781 for your comments. I've removed the use of the walrus operator, even though `pyauditor` does not work with Python 3.6 as @maxfischer2781 pointed out (I've created a ticket for that: #267).
I've also introduced a custom exception which is thrown if no record can be generated.

stefan-k replied:
> Scratch 3.6. I only checked the pyproject.toml but the Maturin build backend already dropped support for 3.6. 

We're using an older version of PyO3 in order to be able to build for Python 3.6 (since TARDIS requires this too). Therefore I believe it should be possible to support 3.6, but there is still the question whether it is worth the effort for this collector.

rfvc commented on collectors/htcondor/collector.py:
```diff
@@ -0,0 +1,242 @@
+        jobs = self.query_htcondor_history(schedd_name, cluster, proc)
+
+        added, failed = 0, 0
+        for job in jobs[::-1]:
```
Yes, true. Apparently that is also faster... Thanks for the comment!

rfvc commented on collectors/htcondor/utils.py:
```diff
@@ -0,0 +1,103 @@
+            return None
+
+    if "only_if" in config_entry:
+        cond = re.search(
+            config_entry["only_if"]["matches"], job[config_entry["only_if"]["key"]]
+        )
```
That's true. I've added a configuration check in 1e138ed, testing if the config parameters are of the expected structure and type.

QuantumDancer replied:
Great, thanks a lot!

rfvc commented:
Concerning the type annotations, that's certainly true. I've added them (hopefully) everywhere in 23e2820.

QuantumDancer commented:
Hey @rfvc,
I'm just checking in to see if there are any stoppers on your side that you would need help with. Michael told me that you are currently working on the integration tests for this plugin. Just for my overview: Is there anything else missing?
Cheers

rfvc commented:
Hey @QuantumDancer, @dirksammel, I think I have addressed all requested changes, except the integration tests. I'm still not done with those, as I am having some issues with networking/authentication which I am working on. For keeping track of this I have opened #301
Maybe you guys can re-review and we can merge for the moment? 
Cheers

dirksammel commented on collectors/htcondor/collector.py:
```diff
@@ -0,0 +1,256 @@
+
+    def setup_auditor_client(self) -> AuditorClient:
+        """Sets up the auditor client."""
+        builder = AuditorClientBuilder().address(self.config.addr, self.config.port)
```
Could it be useful to add a `timeout`?

rfvc replied:
This is addressed in [573b520](https://github.com/ALU-Schumacher/AUDITOR/pull/264/commits/573b5208826228ac336cab8049d49adb7d306f86).
